xarray/core/groupby.py

```diff
@@ -877,7 +877,22 @@ def reduce_array(ar):
         return self.map(reduce_array, shortcut=shortcut)
 
 
-class DatasetGroupBy(GroupBy, DatasetGroupbyArithmetic):
+class DatasetGroupByCombine(GroupBy):
+    def _combine(self, applied):
+        """Recombine the applied objects like the original."""
+        applied_example, applied = peek_at(applied)
+        coord, dim, positions = self._infer_concat_args(applied_example)
+        combined = concat(applied, dim)
+        combined = _maybe_reorder(combined, dim, positions)
+        # assign coord when the applied function does not return that coord
+        if coord is not None and dim not in applied_example.dims:
+            combined[coord.name] = coord
+        combined = self._maybe_restore_empty_groups(combined)
+        combined = self._maybe_unstack(combined)
+        return combined
+
+
+class DatasetGroupBy(DatasetGroupByCombine, DatasetGroupbyArithmetic):
```

Comment:
Pulled out combine because DatasetGroupBy → DatasetGroupByArithmetic → IncludeReduceMethods overwrites the reductions on `WeighedGroupBy`

Reply:
Hmm... when we implement groupby with numpy_groupies, we'll probably have to get rid of `IncludeReduceMethods`. So maybe this should just wait till that happens?